### PR TITLE
Implement async collect of sys tables iterables

### PR DESCRIPTION
**Pull request by matriv**

### CODE REVIEW

File: sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java [hunk 1]
```diff
@@ -65,32 +66,32 @@ public InformationSchemaIterables(final Schemas schemas,
         sqlFeatures = new SqlFeaturesIterable();
     }
 
-    public Iterable<SchemaInfo> schemas() {
-        return schemas;
+    public CompletableFuture<Iterable<SchemaInfo>> schemas() {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/sources/InformationSchemaIterables.java]: I'd tend to wrap them on the call-site instead - allows for more flexibility in how they're used. E.g. in a test it isn't necessary to unwrap the iterable.

File: sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java [hunk 2]
```diff
@@ -445,7 +448,10 @@ private CrateCollector getShardsCollector(RoutedCollectPhase collectPhase,
 
         return new RowsCollector(
             rowReceiver,
-            Iterables.transform(rows, Buckets.arrayToRowFunction()));
+            CompletableFuture.completedFuture(rows),
+            objects -> Iterables.transform(
+                objects,
+                (Function<Object, Row>) input -> Buckets.arrayToRowFunction().apply((Object[]) input)));
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/sources/ShardCollectSource.java]: I think the `Object[]` cast could be avoided if the Function signature is changed.

File: sql/src/main/java/io/crate/operation/collect/stats/JobsLogs.java [hunk 3]
```diff
@@ -138,20 +138,20 @@ public void operationFinished(@Nullable Integer operationId, @Nullable UUID jobI
         operationContextLogs.add(new OperationContextLog(operationContext, errorMessage));
     }
 
-    public Iterable<JobContext> jobsGetter() {
-        return jobsTable.values();
+    public CompletableFuture<Iterable<JobContext>> jobsGetter() {
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/stats/JobsLogs.java]: same here, I'd tend to wrap them on call-site

**Comment by matriv** [comment]
@mfussenegger fixup pushed

File: sql/src/main/java/io/crate/operation/collect/RowsCollector.java [hunk 4]
```diff
@@ -27,48 +27,65 @@
...
 import javax.annotation.Nullable;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Function;
 
 public class RowsCollector implements CrateCollector {
 
-    private final IterableRowEmitter emitter;
+    private final RowReceiver rowDownstream;
+    private final CompletableFuture<? extends Iterable<?>> future;
+    private final Function<Iterable, Iterable<? extends Row>> transformFunction;
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/RowsCollector.java]: Could the naming be improved a bit? 
Maybe  `transformFunction -> dataIterableToRows`

File: sql/src/main/java/io/crate/operation/collect/RowsCollector.java [hunk 5]
```diff
@@ -27,48 +27,65 @@
...
+            CompletableFuture.completedFuture(ImmutableList.of()),
+            objects -> ImmutableList.of());
     }
 
     public static RowsCollector single(Row row, RowReceiver rowDownstream) {
-        return new RowsCollector(rowDownstream, ImmutableList.of(row));
+        return new RowsCollector(
+            rowDownstream,
+            CompletableFuture.completedFuture(ImmutableList.of(row)),
+            objects -> (Iterable<? extends Row>) objects);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/operation/collect/RowsCollector.java]: Maybe rename `objects` to `rows`?
I'm not sure - somehow the term `rows` could refer to both `Iterable<?>` and `Iterable<Row>` which is a bit confusing I guess.

Maybe distinguish via `dataIterable` and `rowIterable` ?

**Pull request merged**